Import Framework Code

msmk0 opened the pull request:
This imports the [framework code](https://gitlab.cern.ch/acts/acts-framework) into the main repository and integrates it into the build system. The code is mostly imported as-is with only minor fixes to adapt to change in the core library. The code has been reorganized in parts to avoid naming collisions/ confusion and now is stored as follows

```
Examples/
  Algorithm/ # contains some code that was previously in plugins
  Detectors/ # contains also the magnetic fields
  Framework/ # formerly `Core`
  Run/ # formerly `Examples` to clarify that this contains full chain tools
  Io/ # contains HepMC3/Json/Obj code that was previously in plugins
```

Renaming this module to `Examples` was done to clarify the intention. I want to avoid that people think this also defines a production-ready event processing framework.

The standalone code can be activated via additional cmake flags
```
CMAKE_BUILD_EXAMPLES
CMAKE_BUILD_EXAMPLES_DD4HEP
CMAKE_BUILD_EXAMPLES_HEPMC3
...
```
Any standalone component flag activates the overall standalone flag as well. Dependencies between option flags are handled automatically, e.g. `CMAKE_BUILD_EXAMPLES_DD4HEP` actives `CMAKE_BUILD_DD4HEP_PLUGIN`. The standalone TGeo detectors are now always build since the standalone code requires ROOT anyways.

Since there are quite a large number of flags now, I added `CMAKE_BUILD_EVERYTHING` flag that does exactly what it says. Build artifacts are now placed in `<build_dir>/bin` to simplify usage.

While the code still uses the `ACTFW/` header hierarchy and `namespace FW`, I updated the CMake target names to match our convention. All internal libraries are prefixed `ActsExamples...`. Executables are prefixed `Acts{Example,Gen,Sim,Rec}...` as appropriate.

The two examples that previous existed in `Tests/Examples` have been moved to `Examples/Run/Misc`.

Closes #22

paulgessinger commented:
FYI: The labels Infrastructure and New Feature are treated mutually exclusive.

paulgessinger commented on .github/workflows/build.yml:
```diff
@@ -12,89 +12,57 @@ jobs:
+        && cd build
         && cmake ..
         -DCMAKE_BUILD_TYPE=Release
-        -DCMAKE_CXX_STANDARD=17
```
Do you specify the standard version explicitly somewhere else?

msmk0 replied:
It is already specified per target

msmk0 replied:
Btw, I still need to get the CI build running. The build configuration might still change.

msmk0 commented:
> FYI: The labels Infrastructure and New Feature are treated mutually exclusive.

Thanks, fixed it.

paulgessinger commented:
Can you rename the options from `*_STANDALONE`? I think it's pretty opaque what that means / doesn't refer to it being the testing framework.

paulgessinger commented on README.md:
```diff
@@ -45,16 +45,20 @@ If you find a bug, have a feature request, or want to contribute to Acts, please
 
-This software contains a copy of the `JSON for Modern C++`-library by Niels Lohmann licensed under the MIT License.
+The Acts project contains copies of the following external packages:
 
-The Acts project is based on the ATLAS tracking software. A list of contributors
-to the ATLAS tracking repository can be found <a href="http://acts.web.cern.ch/ACTS/ATLAS_authors.html">here</a>.
+*   [dfelibs](https:://gitlab.cern.ch/msmk/dfelibs) by Moritz Kiehn licensed
+    under the MIT license.
+*   [gcovr](http://gcovr.com) licensed under the 3-Clause BSD license.
+*   [JSON for Modern C++](https://github.com/nlohmann/json) by Niels Lohmann
+    licensed under the MIT License.
```
I think this isn't true anymore. We don't ship `nlohmann::json` anymore (CMake downloads it), and `dfelibs` is a submodule. The bundled `gcovr` is not used anymore, so we can throw it out if it's still there.

Can you remove the entire section?

paulgessinger replied:
Nevermind, we changed that.

paulgessinger replied:
I'm still reaaaally uncomfortable with having these commited inside our repository thought :/

msmk0 replied:
Both `nlohmann::json` and `dfelibs` are shipped as source copies in `thirdparty` in order to not have submodules anymore (that is one of the reason for going towards a monorepo). We decided in https://gitlab.cern.ch/acts/acts-core/-/merge_requests/733 to ship `nlohmann::json` by copy for various reasons.

msmk0 commented:
> Can you rename the options from `*_STANDALONE`? I think it's pretty opaque what that means / doesn't refer to it being the testing framework.

I would prefer to leave the naming as `Standalone` for the reasons discussed in the PR description. To expand a bit on this: the purpose of this code is to run full standalone reconstruction chains both to develop the code, run computation performance studies, and perform physics performance studies.

We could consider them as just another set of tests and place them under the `Tests` hierarchy. But, the majority of the functionality is not and/or can not be run as automated tests. The code is also quite large and does way more than just test some part of the core code. That is enough reason to organize it as separate from the tests.

The fact that we use our own event processing framework internally is an implementation detail at this point. Thus, the change to `Standalone`.

paulgessinger commented:
> Renaming this module to Standalone was done to clarify the intention. I want to avoid that people think this also defines a production-ready event processing framework.

Actually, I think the name "Standalone" achieves the exact opposite of what you state the purpose is. But ok, I guess we don't have to agree on this.

HadrienG2 commented:
If an extra opinion could be helpful in resolving the disagreement, I personally have no idea what "Standalone" means, and would prefer another terminology like "full chain", "usage examples" or "integration tests".

asalzburger commented:
If `Standalone` is disputed, we could rename it to `Examples` ?

msmk0 commented:
I moved the code from `Standalone` to `Examples` as discussed. The PR description has been updated accordingly.

I added one more change: executables can now be found in `<build_dir>/bin` instead of e.g. `<build_dir>/Examples/Run/MagnetifField/ActsExampleMagneticField...`.

asalzburger commented:
Thank you @msmk0 - I think that's a good compromise.

asalzburger commented:
Why do we now have `Builds / macos` and `build_macos` e.g. ?

asalzburger commented:
Since I merged #98 in,  here is a super simple recipe to adapt the `Examples` code to the changes:

`#include "Acts/Propagation/detail/StandardAborter.hpp"` changed to  `"Acts/Propagation/StandardAborter.hpp" `

Consequence:
- `detail::PathLimitReached` -> `PathLimitReached`
- `detail::EndOfWorldReached` -> `EndOfWorldReached`
- `detail::SurfaceReached` -> `SurfaceReached`

The same happened with the `DebugOutputActor`.

msmk0 commented:
Just a quick update: everything seems to build now with the updated [container images](https://github.com/acts-project/machines/pull/2). Only the macOS build fails now due to an CMake error during configuration. The `Threads` library can not be found. This is a bit surprising since this is a standard component that **should** be there and the macOS build was previously successful. A Google search suggests that this might be a bug in the most recent XCode versions and a downgrade to an earlier version might help.

I do not have access to a macOS machine. Could someone who uses macOS try to reproduce the issue locally and have a look at possible solutions?

msmk0 commented:
I have switched to the updated container images from the CERN registry. All builds compile know, including the macos one (thanks to @paulgessinger). From my side this is ready to be merged and I removed the WIP flag.

There seems to be still some issue with the coverage reporting, but I can not see any log files or errors. There also appear some required checks that do not exists anymore/ have been renamed to match e.g. the lcg name. I'm not sure if there are any other files that need modification for this. @paulgessinger Do you have any idea?

paulgessinger commented:
The required checks are set in the Github settings interface. I'll switch that over when we merge this, and override it for this PR.

msmk0 commented:
Rebased and still ready to be merged.

asalzburger commented:
Will check for the build and then merge.

I will then open an Issue for @Corentin-Allaire and @asalzburger  to define a set of tests that are defined as new UnitTests with the Examples being built.

A few quick question, you can just check if true
- [ ] are we counting the Examples for code coverage ? 
- [ ] are we building the Examples for every build ?

paulgessinger commented:
I think I would prefer to merge #102 before merging this one, @asalzburger.

msmk0 commented:
All builds use the new `ACTS_BUILD_EVERYTHING` flag that includes the examples build. `Examples` should be excluded from the coverage but @paulgessinger should verify this.

This PR only has minimal changes to the Core and should therefore be orthogonal to #102. A potential rebase after merging this first should only be a technical matter. Is there another reason for wanting to postpone this?

paulgessinger commented:
I think it's probably easier for you to rebase this on top of #102 because you're familiar with the technical changes in this PR that might potentially affect #102's code than doing it the other way around.

Corentin-Allaire commented:
I tried to run the mapping to check everything was fine and I noticed the condition in the CMakeList for the mapping were not update. 
In Examples/Algorithm/MaterialMapping/CMakeList.txt and Examples/Run/MaterialMapping/CMakeList.txt USE_DD4HEP and USE_GEANT4 need to be replace by ACTS_BUILD_EXAMPLES_DD4HEP and ACTS_BUILD_EXAMPLES_GEANT4.

Corentin-Allaire commented on Examples/Algorithms/MaterialMapping/src/GeantinoRecording.cpp:
```diff
@@ -0,0 +1,77 @@
+
+  /// Now set up the Geant4 simulation
+  m_runManager->SetUserInitialization(new FTFP_BERT);
+  m_runManager->SetUserAction(new FW::Geant4::MMPrimaryGeneratorAction(
+      "mu-", 1000., m_cfg.seed1, m_cfg.seed2));
```
Going through the material mapping I noticed that in the GeantinoRecording Algorithm the geantino have been replace with muons. I guess you didn't do that and it was done in a previous MR that I didn't look at but could you fix it ?

Corentin-Allaire replied:
It is also extremely slow compare to what it use to be. I will investigate

msmk0 replied:
This was changed in a PR by Stany to be able to do Geant4/Fatras comparison. I can switch this back for now. I the long run, we need to be able to run Geant4 with arbitrary particle input.

Corentin-Allaire replied:
Isn't that the purpose of the ActsGenParticleGun Algorithm ?

I also notice that in that PR she added the recording of track step which make the reording slower by a factor 60, maybe all that should be in another algorithm (or at least have an option to turn it off) so that the geantino recording only does the geantino recording.

I will remove that part of the code locally for now and continue to try the mapping chain.

msmk0 replied:
> Isn't that the purpose of the ActsGenParticleGun Algorithm ?

Yes, but we need to adapt the Geant4 algorithms to take the generated particles as inputs; instead of relying on the Geant4 internal particle gun.

> I also notice that in that PR she added the recording of track step which make the reording slower by a factor 60, maybe all that should be in another algorithm (or at least have an option to turn it off) so that the geantino recording only does the geantino recording.
> 
> I will remove that part of the code locally for now and continue to try the mapping chain.

In that case, a separate algorithm is probably the better choice. I will open an issue.

msmk0 replied:
Tracked in #111 now.

Corentin-Allaire replied:
Thanks !

msmk0 commented:
> I think it's probably easier for you to rebase this on top of #102 because you're familiar with the technical changes in this PR that might potentially affect #102's code than doing it the other way around.

asalzburger commented:
What's the conclusion on this?

Shall we put that in before the combinatorial Kalman filter  or the other way round?

Corentin-Allaire commented:
I went thought the full material mapping and everything seems fine (except for what is mentioned in #111).
But I think the script directory (in which the macro to draw the validation plots are) is missing.

asalzburger commented:
> I went thought the full material mapping and everything seems fine (except for what is mentioned in #111).
> But I think the script directory (in which the macro to draw the validation plots are) is missing.


Thanks, I will try to fix the navigation in the ITk then.

msmk0 commented:
> What's the conclusion on this?
> 
> Shall we put that in before the combinatorial Kalman filter or the other way round?

I will add the fixes requested by @Corentin-Allaire and then we can merge. I would suggest to just merge whatever is ready first. But there is no reason why I should go in in any particular order.

asalzburger commented:
Ouch, coverage now takes 92m ?

msmk0 commented:
@Corentin-Allaire Missing `Examples/Scripts` have been added. The CMake flags should be used correctly. The geantino recording should work as before.

Corentin-Allaire commented:
Just tested and now the Geantino recoding works great !

asalzburger commented:
> Just tested and now the Geantino recoding works great !

Is that with the ODD or on top of my irk-navigation branch ?

Corentin-Allaire commented:
> > Just tested and now the Geantino recoding works great !
> 
> Is that with the ODD or on top of my irk-navigation branch ?

With the ODD, I was testing that the slowness that was caused by the recording of the track step was not here anymore.

msmk0 commented:
This could be merged, but the coverage report consistently fails now. Not sure if we want to investigate this first or override it for now?

asalzburger commented:
I think we merge this.

msmk0 commented on Examples/Algorithms/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+add_subdirectory_if(DD4hepG4 ACTS_BUILD_EXAMPLES_DD4HEP AND ACTS_BUILD_EXAMPLES_GEANT4)
```
Yes, because DD4hep and HepMC3 are not generally available in the CI builds. DD4hep is missing from the Ubuntu 20.04 image but will be added again (#129). HepMC3 is not available in LCG95 and it can only be dropped from the `EVERYTHING` flag once we drop the LCG95 support.
